protobuf/module_msg.proto

```diff
@@ -6,385 +6,869 @@ package bess.pb;
+    string internal_addr_block = 1; /// Internal IP block in CIDR.
+    string external_addr_block = 2; /// External IP block in CIDR.
   }
-  repeated Rule rules = 1;
+  repeated Rule rules = 1; /// A list of rules for rewriting
 }
 
+/**
+ * This module is used for testing purposes.
+ */
 message NoOpArg {
 }
 
+/**
+ * The PortInc module connects a physical or virtual port and releases
+ * packets from it. PortInc does not support multiqueueing.
+ * For details on how to configure PortInc using DPDK, virtual ports,
+ * or libpcap, see the sidebar in the wiki.
+ *
+ * __Input Gates__: 0
+ * __Output Gates__: 1
+ */
 message PortIncArg {
-  string port = 1;
-  int64 burst = 2;
-  bool prefetch = 3;
-}
-
+  string port = 1; /// The portname to connect to.
+  int64 burst = 2; /// The number of packets per batch to release.
+  bool prefetch = 3; /// Whether or not to prefetch packets from the port.
+}
+
+/**
+ * The PortOut module connects to a physical or virtual port and pushes
+ * packets to it. For details on how to configure PortOut with DPDK,
+ * virtual ports, libpcap, etc, see the sidebar in the wiki.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 0
+ */
 message PortOutArg {
-  string port = 1;
-}
-
+  string port = 1; /// The portname to connect to.
+}
+
+/**
+ * The QueueInc produces input packets from a physical or virtual port.
+ * Unlike PortInc, it supports multiqueue ports.
+ * For details on how to configure QueueInc with DPDK, virtualports,
+ * libpcap, etc, see the sidebar in the wiki.
+ *
+ * __Input Gates__: 0
+ * __Output Gates__: 1
+ */
 message QueueIncArg {
-  string port = 1;
-  uint64 qid = 2;
-  int64 burst = 3;
-  bool prefetch = 4;
-}
-
+  string port = 1; /// The portname to connect to (read from).
+  uint64 qid = 2; /// The queue on that port to read from.
+  int64 burst = 3; /// The maximum number of packets to release in a batch.
+  bool prefetch = 4; /// Ehrthrt ot not to prefetch packets from the port.
+}
+
+/**
+ * The QueueOut module releases packets to a physical or virtual port.
+ * Unlike PortOut, it supports multiqueue ports.
+ * For details on how to configure QueueOut with DPDK, virtualports,
+ * libpcap, etc, see the sidebar in the wiki.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 0
+ */
 message QueueOutArg {
-  string port = 1;
-  uint64 qid = 2;
+  string port = 1; /// The portname to connect to.
+  uint64 qid = 2; /// The queue on that port to write out to.
 }
 
+/**
+ * The Queue module implements a simple packet queue.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 1
+ */
 message QueueArg {
-  uint64 size = 1;
-  int64 burst = 2;
-  bool prefetch = 3;
-}
-
+  uint64 size = 1; /// The maximum number of packets to store in the queue.
+  int64 burst = 2; /// The maximum number of packets to release from the queue in one batch.
+  bool prefetch = 3; /// Whether or not to prefetch packet data for the next batch after releasing. SANGJIN WHAT IS THE PURPOSE OF THIS?
```

Comment:
When `prefetch` is enabled, the module will perform CPU prefetch on the first 64B of each packet onto CPU L1 cache. This is just an optimization knob; whether it is enabled does not affect functionality. This option can be useful if it is already known that the packet data is cache-cold or is in another core's private cache.

Default value is `false`.

All Queue, PortInc, and QueueInc share the same semantics for this `prefetch` parameter.